File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/components/UseSmileIDSampleOptionRow.kt
package com.usesmileid.sampleapps.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.sp
import com.smileid.designsystem.SmileDimens
import com.usesmileid.sampleapps.ui.theme.UseSmileIDSampleTheme

/** One row in the country or ID-type picker; country rows lead with a flag. Selected takes a pale fill as well as a check. */
@Composable
fun UseSmileIDSampleOptionRow(
    label: String,
    selected: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    leadingText: String? = null,
    testId: String? = null,
) {
    val colors = UseSmileIDSampleTheme.colors
    Surface(
        modifier = modifier
            .fillMaxWidth()
            .selectable(selected = selected, role = Role.RadioButton, onClick = onClick)
            .tagged(testId),
        shape = RoundedCornerShape(SmileDimens.radiusField),
        // An unselected row is transparent over the sheet; only the selected one takes a fill.
        color = if (selected) UseSmileIDSampleTheme.colors.surfaceMuted else Color.Transparent,
    ) {
        Row(
            modifier = Modifier
                .defaultMinSize(minHeight = SmileDimens.sizeControlMd)
                .padding(horizontal = SmileDimens.spacingSm, vertical = SmileDimens.spacingXs),
            horizontalArrangement = Arrangement.spacedBy(SmileDimens.spacingSm),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (leadingText != null) {
                Text(text = leadingText, style = UseSmileIDSampleTheme.type.textStyleBody.copy(fontSize = FLAG_SIZE))
            }
            Text(
                text = label,
                style = UseSmileIDSampleTheme.type.textStyleBodyStrong.copy(fontSize = OPTION_LABEL_SIZE),
                color = colors.textTitle,
                modifier = Modifier.weight(1f),
            )
            if (selected) {
                Box(modifier = Modifier.size(SmileDimens.sizeIconMd), contentAlignment = Alignment.Center) {
                    CheckGlyph(tint = colors.primary)
                }
            }
        }
    }
}

private val FLAG_SIZE = 19.sp
private val OPTION_LABEL_SIZE = 14.sp

File: flutter/sample_ui/lib/src/components/use_smileid_sample_option_row.dart
import 'package:flutter/material.dart';

import '../theme/use_smileid_sample_colors.dart';
import '../theme/use_smileid_sample_theme.dart';
import '../theme/use_smileid_sample_typography.dart';
import '../tokens/smile_tokens.dart';
import 'use_smileid_sample_glyphs.dart';

/// One row in the country or ID-type picker; country rows lead with a flag.
class UseSmileIDSampleOptionRow extends StatelessWidget {
  /// [leadingText] is the flag emoji, which is text rather than an asset and so needs no icon set.
  const UseSmileIDSampleOptionRow({
    required this.label,
    required this.selected,
    required this.onTap,
    this.leadingText,
    this.testId,
    super.key,
  });

  /// The country or ID-type name.
  final String label;

  /// Whether this option is chosen.
  final bool selected;

  /// What a tap does.
  final VoidCallback onTap;

  /// The country's flag, absent on an ID-type row.
  final String? leadingText;

  /// The `sample_*` id the sheet supplies.
  final String? testId;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(
      context,
    );
    return Semantics(
      identifier: testId,
      selected: selected,
      child: Material(
        color: selected ? colors.surfaceTile : Colors.transparent,
        borderRadius: BorderRadius.circular(SmileDimens.radiusField),
        child: InkWell(
          onTap: onTap,
          borderRadius: BorderRadius.circular(SmileDimens.radiusField),
          child: ConstrainedBox(
            constraints: const BoxConstraints(
              minHeight: SmileDimens.sizeControlMd,
            ),
            child: Padding(
              padding: const EdgeInsets.symmetric(
                horizontal: SmileDimens.spacingSm,
                vertical: SmileDimens.spacingXs,
              ),
              child: Row(
                children: <Widget>[
                  if (leadingText != null) ...<Widget>[
                    Text(
                      leadingText!,
                      style: UseSmileIDSampleType.textStyleBody.copyWith(
                        fontSize: _flagSize,
                      ),
                    ),
                    const SizedBox(width: SmileDimens.spacingSm),
                  ],
                  Expanded(
                    child: Text(
                      label,
                      style: UseSmileIDSampleType.textStyleBodyStrong.copyWith(
                        fontSize: _optionLabelSize,
                        color: colors.textTitle,
                      ),
                    ),
                  ),
                  if (selected) ...<Widget>[
                    const SizedBox(width: SmileDimens.spacingSm),
                    SizedBox(
                      width: SmileDimens.sizeIconMd,
                      height: SmileDimens.sizeIconMd,
                      child: Center(
                        child: UseSmileIDSampleGlyphs.check(colors.primary),
                      ),
                    ),
                  ],
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}

/// The flag emoji's run, which grows with the font scale.
const double _flagSize = 19;

/// Body Strong at its real 14.
const double _optionLabelSize = 14;
